cmake: Add option to generate precomputed files

hebasto opened the pull request:
Using the CMake-based build system enables smoother integration with parent projects by providing both CMake package configuration files and the traditional `libsecp256k1.pc`.

However, packagers on some systems, such as [Gentoo](https://github.com/bitcoin-core/secp256k1/blob/f9a944ff2dbcd7ff1829880636f861caf0a91392/Makefile.am#L83) and [Guix](https://codeberg.org/guix/guix/pulls/2566#issuecomment-7360201), insist on recreating the precomputed files, which is not currently supported by the CMake-based build system.

This PR addresses the issue by introducing a new `SECP256K1_BUILD_PRECOMPUTED` option (`OFF` by default) that generates `precomputed_ecmult.c` and `precomputed_ecmult_gen.c` in the build tree. Those files are then compiled instead of the ones checked into the source tree.

Here is an example:
```
$ cmake -B build -G Ninja -DSECP256K1_BUILD_PRECOMPUTED=ON
$ ninja -C build src/precomputed_ecmult.c src/precomputed_ecmult_gen.c | cat
ninja: Entering directory `build'
[1/6] Building C object src/CMakeFiles/secp256k1_precompute_ecmult.dir/precompute_ecmult.c.o
[2/6] Building C object src/CMakeFiles/secp256k1_precompute_ecmult_gen.dir/precompute_ecmult_gen.c.o
[3/6] Linking C executable secp256k1_precompute_ecmult_gen
[4/6] Linking C executable secp256k1_precompute_ecmult
[5/6] Generating precomputed_ecmult_gen.c
[6/6] Generating precomputed_ecmult.c
```

When configuring with CMake 3.31 or newer, the built-in `codegen` target can be used to generate those files:
```
$ cmake -B build -G Ninja -DSECP256K1_BUILD_PRECOMPUTED=ON
$ cmake --build build -t codegen | cat
[1/6] Building C object src/CMakeFiles/secp256k1_precompute_ecmult.dir/precompute_ecmult.c.o
[2/6] Building C object src/CMakeFiles/secp256k1_precompute_ecmult_gen.dir/precompute_ecmult_gen.c.o
[3/6] Linking C executable secp256k1_precompute_ecmult_gen
[4/6] Linking C executable secp256k1_precompute_ecmult
[5/6] Generating precomputed_ecmult_gen.c
[6/6] Generating precomputed_ecmult.c
```

kevkevinpal commented:
are we able to do this via `./configure`? 

or is this meant to be only done for CMake?

hebasto commented:
> are we able to do this via `./configure`?

Yes. Run `make clean-precomp && make precomp`.

real-or-random commented on src/CMakeLists.txt:
```diff
@@ -45,10 +45,41 @@ if(SECP256K1_ENABLE_MODULE_ECDH)
+  if(CMAKE_C_COMPILER_ID STREQUAL "MSVC")
+    # Disable warning C4334 "'operator': result of 32-bit shift implicitly converted to 64 bits (was 64-bit shift intended?)".
+    target_compile_options(secp256k1_precompute_ecmult PRIVATE /wd4334)
+  endif()
```
In which line does this warning appear? Could we fix it in the code instead?

real-or-random reviewed (commented):
Concept ACK

hebasto commented on src/CMakeLists.txt:
```diff
@@ -45,10 +45,41 @@ if(SECP256K1_ENABLE_MODULE_ECDH)
+    COMMAND secp256k1_precompute_ecmult_gen
+    DEPENDS secp256k1_precompute_ecmult secp256k1_precompute_ecmult_gen
+    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+  )
```
Thanks! Reworked.

theStack commented:
Concept ACK

Tested quickly that 

```
$ cmake -B build -DSECP256K1_ECMULT_WINDOW_SIZE=16 -DSECP256K1_BUILD_PRECOMPUTED=ON && cmake --build build
```
succeeds on the PR branch as expected, while on master
```
$ cmake -B build -DSECP256K1_ECMULT_WINDOW_SIZE=16 && cmake --build build
```
fails (`error: #error configuration mismatch, invalid ECMULT_WINDOW_SIZE`).

The generated files `precomputed_ecmult{,gen}.c` files are created within the build directory. Given the out-of-tree build nature of CMake, I guess this is intended and there are no plans to support directly modifying the source tree? (Which is no big deal, developers can simply do a `$ cp ./build/src/precomputed_ecmult*.c ./src/` after manually or just re-run the precompute binaries).

hebasto commented:
> Given the out-of-tree build nature of CMake, I guess this is intended and there are no plans to support directly modifying the source tree?

That's correct. We have to assume the source tree might be strictly read-only.

theStack reviewed (approved):
Light ACK 8b58c55e3d4b5e18a6bb626c5b55548b832965eb

I'm not too familiar with the ins and outs of CMake, but the changes look good and I tested the new build setting both with higher ecmult window sizes (as [shown above](https://github.com/bitcoin-core/secp256k1/pull/1791#issuecomment-4632407609)) and with different ecmult_gen `CONFIGS` table values (taken from https://github.com/theStack/secp256k1/commit/3421153be6ae8d8d88406d518c9e66ef4d49fb0a), and everything seems to work as expected.

// EDIT: I guess this is worth a release note

real-or-random commented on CMakeLists.txt:
```diff
@@ -141,6 +141,8 @@ option(SECP256K1_BUILD_EXHAUSTIVE_TESTS "Build exhaustive tests." ON)
 option(SECP256K1_BUILD_CTIME_TESTS "Build constant-time tests." ${SECP256K1_VALGRIND})
 option(SECP256K1_BUILD_EXAMPLES "Build examples." OFF)
 
+option(SECP256K1_BUILD_PRECOMPUTED "Recreate the precomputed files." OFF)
```
nit: Do you think it will add clarity to expand this to "Recreate the precomputed files in the build tree"?

It's kind of implied that a CMake build creates stuff only in the build tree. But @theStack felt the need to point this out, so this may be surprising to users who don't really think about carefully (which includes almost everyone when it comes to build systems, I guess).

hebasto replied:
Thanks! [Amended](https://github.com/bitcoin-core/secp256k1/pull/1791#issuecomment-4659102568).

hebasto commented:
The most recent feedback has been addressed.

theStack reviewed (approved):
re-ACK 8460c5902cf83d3b83e22021c7d73d6991347464

Verified that the only change since my last review was the more detailed build option description (+"in the build tree"), which I agree makes sense.

real-or-random commented:
> fails (`error: #error configuration mismatch, invalid ECMULT_WINDOW_SIZE`).

The truth is that it also outputs `Try deleting precomputed_ecmult.c before the build.` And this is not true for the approach in this PR. With Automake, we construct real dependencies, i.e., the main artifact depends on the pregenerated files. Here, this is just a build option.

I think either approach is fine. I guess even mixing them is fine. Maybe we should just change the error message for now? "If this is an autotools build, try deleting precomputed_ecmult.c before the build. If this is a CMake build, pass -DSECP256K1_BUILD_PRECOMPUTED=ON to cmake."

---

Further thoughts:

I'm wondering if I like the "build option" approach as in this PR more. It's easier for those users who need it, and the code is simpler. And even for contributors working on changes to the generation binaries, they can just set -DSECP256K1_BUILD_PRECOMPUTED=ON in their cache.

If others agree, this raises the question of whether we want to get rid of that stuff in autotools. It never worked 100%. And it seems that all users who want this feature use CMake. (@kevkevinpal Why did you ask about autotools? Do you need this feature in autotools?)

real-or-random commented on CMakePresets.json:
```diff
@@ -7,7 +7,8 @@
         "SECP256K1_ENABLE_MODULE_RECOVERY": "ON",
-        "SECP256K1_BUILD_EXAMPLES": "ON"
+        "SECP256K1_BUILD_EXAMPLES": "ON",
+        "SECP256K1_BUILD_PRECOMPUTED": "ON"
```
I'm not entirely sure if we want this in dev mode. In autotools, we don't have it in dev mode. 

It's something a dev needs when working on the source for the precomputation binaries. But on the other hand, this is rare and this thing increases compile times. Maybe leave it off for now?

hebasto replied:
Thanks! [Amended](https://github.com/bitcoin-core/secp256k1/pull/1791#issuecomment-4661608463).

hebasto replied:
> Maybe leave it off for now?

That raises the question: which CI job should use `-DSECP256K1_BUILD_PRECOMPUTED=ON`?

real-or-random replied:
My first thought is "Why not all?"

real-or-random reviewed (commented):
Consider rebasing this. The branch is way behind (for our standards ;)).

hebasto commented:
> Consider rebasing this. The branch is way behind (for our standards ;)).

Rebased. The recent feedback has been addressed.

> Maybe we should just change the error message for now? "If this is an autotools build, try deleting precomputed_ecmult.c before the build. If this is a CMake build, pass -DSECP256K1_BUILD_PRECOMPUTED=ON to cmake."

Done.

hebasto commented on cmake/Precompute.cmake:
```diff
@@ -0,0 +1,13 @@
+  set_target_properties(secp256k1_precompute_${base_name} PROPERTIES
+    COMPILE_DEFINITIONS VERIFY
+    RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}
+  )
+  add_custom_command(
+    OUTPUT ${PROJECT_BINARY_DIR}/src/precomputed_${base_name}.c
+    COMMAND secp256k1_precompute_${base_name}
+    DEPENDS secp256k1_precompute_${base_name}
+    WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+  )
```
`secp256k1_precompute_${base_name}` is an executable target name.

From CMake [docs](https://cmake.org/cmake/help/latest/command/add_custom_command.html):
> If `COMMAND` specifies an executable target name (created by the [`add_executable()`](https://cmake.org/cmake/help/latest/command/add_executable.html) command), it will automatically be replaced by the location of the executable created at build time...

hebasto commented on src/CMakeLists.txt:
```diff
@@ -45,10 +45,22 @@ if(SECP256K1_ENABLE_MODULE_ECDH)
-  precomputed_ecmult_gen.c
-)
+add_library(secp256k1_precomputed OBJECT EXCLUDE_FROM_ALL)
+if(SECP256K1_BUILD_PRECOMPUTED)
+  include(Precompute)
+  precompute(ecmult)
+  precompute(ecmult_gen)
```
Yes. The current CMake implementation doesn't support cross-compiling. And neither does Autotools.

FWIW, I have a branch that introduces this support, but I'm not sure whether the extra complexity is justified.

real-or-random replied:
> FWIW, I have a branch that introduces this support, but I'm not sure whether the extra complexity is justified.

Ah please don't. :D Years ago, when we always ran the precomputation during the build (no precomputed files in repo), we had support for this in Autotools. But handling the different compilers (one for host and one for target) was a nightmare for users who wanted to cross-compile and for us to maintain. It's one of the reasons why added the precomputed files to the repo. 

I think if we really want to support this during cross-compilation, we'd rather rewrite the precomputation helpers in python or sage.

hebasto replied:
> Ah please don't. :D Years ago, when we always ran the precomputation during the build (no precomputed files in repo), we had support for this in Autotools.

TIL. Not arguing, but CMake has built-in ways to achieve this.

real-or-random replied:
> CMake has built-in ways to achieve this.

autoconf has this here: https://www.gnu.org/software/autoconf-archive/ax_prog_cc_for_build.html (not exactly built-in but that's the standard way on autoconf)

I totally believe that the CMake way of doing this is much cleaner, but I think part of the problem is on a conceptual level. Now you get effectively two build configs with all their complexity. You have separate CFLAGS, CPPFLAGS, LDFLAGS, which raises all kinds of questions, e.g., which warning options do you want to set on the compiler for the build machine? Maybe the compiler is entirely different, e.g., you have a new clang for the build machine but an old gcc for the target, etc... And some of this complexity is necessarily pushed down to the user.

Just assuming the existence of Python or similar is probably simpler. Back then we had already considered using Python instead, but approach to add the precomputed files to the repo was just less work, so we went with this instead.

real-or-random replied:
But do you think some error message like suggested by the LLM makes sense for now?

hebasto replied:
> But do you think some error message like suggested by the LLM makes sense for now?

Done.

kevkevinpal commented:
> If others agree, this raises the question of whether we want to get rid of that stuff in autotools. It never worked 100%. And it seems that all users who want this feature use CMake. (@kevkevinpal Why did you ask about autotools? Do you need this feature in autotools?)  

Trying to remember why but I think was just trying to understand why it wasn't available. If anything I agree with you and also prefer CMake.

real-or-random commented:
> > Maybe we should just change the error message for now? "If this is an autotools build, try deleting precomputed_ecmult.c before the build. If this is a CMake build, pass -DSECP256K1_BUILD_PRECOMPUTED=ON to cmake."
> 
> Done.

We have the same message in precomputed_ecmult_gen.c. Sorry, I assumed that's implied. I should have mentioned this explicitly.

real-or-random commented:
For autotools builds on CI, we have `git diff` that checks the files in the repo are not outdated. We could do the same on CI for CMake, i.e., copy the files back to the source tree and run `git diff --exit-code`.

https://github.com/bitcoin-core/secp256k1/blob/0f4a7e6bf9d971addb2b851df7cd2777fc62b212/ci/ci.sh#L138-L147

hebasto commented:
> > > Maybe we should just change the error message for now? "If this is an autotools build, try deleting precomputed_ecmult.c before the build. If this is a CMake build, pass -DSECP256K1_BUILD_PRECOMPUTED=ON to cmake."
> > 
> > 
> > Done.
> 
> We have the same message in precomputed_ecmult_gen.c. Sorry, I assumed that's implied. I should have mentioned this explicitly.

Of course, I've regenerated `precomputed_ecmult_gen.c` in the source tree :)

real-or-random commented:
> > > > Maybe we should just change the error message for now? "If this is an autotools build, try deleting precomputed_ecmult.c before the build. If this is a CMake build, pass -DSECP256K1_BUILD_PRECOMPUTED=ON to cmake."
> > > 
> > > 
> > > Done.
> > 
> > 
> > We have the same message in precomputed_ecmult_gen.c. Sorry, I assumed that's implied. I should have mentioned this explicitly.
> 
> Of course, I've regenerated `precomputed_ecmult_gen.c` in the source tree :)

Oops, we're talking past each other. Sorry, I meant to write `precompute_ecmult_gen.c` (without the `d`). 

So you've changed `precompute_ecmult.c` and `precomputed_ecmult.c` but not `precompute_ecmult_gen.c` and `precomputed_ecmult_gen.c`.

hebasto commented:
Reworked per recent feedback. Added [`CODEGEN`](https://cmake.org/cmake/help/latest/policy/CMP0171.html) support for modern CMake.

The PR description has been updated.

> So you've changed `precompute_ecmult.c` and `precomputed_ecmult.c` but not `precompute_ecmult_gen.c` and `precomputed_ecmult_gen.c`.

My bad. Fixed.

> For autotools builds on CI, we have `git diff` that checks the files in the repo are not outdated. We could do the same on CI for CMake, i.e., copy the files back to the source tree and run `git diff --exit-code`.

Done.

real-or-random commented on CMakePresets.json:
```diff
@@ -7,7 +7,8 @@
         "SECP256K1_ENABLE_MODULE_RECOVERY": "ON",
-        "SECP256K1_BUILD_EXAMPLES": "ON"
+        "SECP256K1_BUILD_EXAMPLES": "ON",
+        "SECP256K1_BUILD_PRECOMPUTED": "OFF"
```
nit: Why not omit this line entirely?

hebasto replied:
Thanks! Dropped.

real-or-random commented on .github/workflows/ci.yml:
```diff
@@ -538,6 +538,10 @@ jobs:
     runs-on: windows-2022
+    defaults:
+      run:
+        # Enforce fail-fast behavior for PowerShell.
+        shell: pwsh -Command "$PSNativeCommandUseErrorActionPreference = $true; $ErrorActionPreference = 'Stop'; & '{0}'"
```
Shouldn't this be implied according to [the docs](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#exit-codes-and-error-action-preference)?

hebasto replied:
> Shouldn't this be implied according to [the docs](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#exit-codes-and-error-action-preference)?

The docs state: "Fail-fast behavior _when possible_." Effectively, it works for cmdlets only. Non-zero exit codes from native commands are ignored unless `$PSNativeCommandUseErrorActionPreference` is also set.

real-or-random replied:
Sigh, why can't we just have nice things...

hebasto committed:
cmake: Add option to generate precomputed files

Add a new `SECP256K1_BUILD_PRECOMPUTED` option (`OFF` by default) that
generates `precomputed_ecmult.c` and `precomputed_ecmult_gen.c` in the
build tree. Those files are then compiled instead of the ones checked
into the source tree.

When configuring with CMake 3.31 or newer, the built-in `codegen` target
can be used to generate those files.

hebasto committed:
ci: Add `-DSECP256K1_BUILD_PRECOMPUTED=ON` to CMake jobs

hebasto commented:
Rebased on top of the recent CI changes.
